ykman/otp.py

```diff
@@ -123,6 +132,56 @@ def program_otp(self, slot, key, fixed, uid, append_cr=True):
         finally:
             ykpers.ykp_free_config(cfg)
 
+    def prepare_upload_key(self, key, public_id, private_id):
+        serial_ref = c_uint()
+        if ykpers.yk_get_serial(self._dev, 0, 0, byref(serial_ref)):
+            serial = serial_ref.value
+        else:
+            serial = 0
+
+        modhex_public_id = modhex_encode(public_id)
+        data = {
+            'aes_key': binascii.b2a_hex(key).decode('utf-8'),
+            'serial': serial,
+            'public_id': modhex_public_id,
+            'private_id': binascii.b2a_hex(private_id).decode('utf-8'),
+        }
+
+        httpconn = http.client.HTTPSConnection(UPLOAD_HOST, timeout=1)
+        try:
+            httpconn.request('POST', UPLOAD_PATH,
+                             body=json.dumps(data, indent=False, sort_keys=True)
+                             .encode('utf-8'),
+                             headers={
+                                 'Content-Type': 'application/json',
+                                 'User-Agent':
+                                 'python-yubikey-manager/' + __version__,
+                             })
+        except Exception as e:
+            logger.error('Failed to connect to %s', UPLOAD_HOST, exc_info=e)
+            return {'success': False, 'error': 'connection_failed'}
+
+        resp = httpconn.getresponse()
+        if resp.status == 200:
+            url = json.loads(resp.read().decode('utf-8'))['finish_url']
+            return {'success': True, 'url': url}
```

Comment:
We don't typically return these kind of result-objects from the library, I would prefer to return only the url and raise an exception on any failure.